flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncCollectorBuffer.java

```diff
@@ -0,0 +1,453 @@
+			TimestampedCollector<OUT> collector,
+			Object lock,
+			AsyncWaitOperator operator) {
+		Preconditions.checkArgument(bufferSize > 0, "Future buffer size should be greater than 0.");
+		Preconditions.checkNotNull(output, "Output should not be NULL.");
+		Preconditions.checkNotNull(collector, "TimestampedCollector should not be NULL.");
+		Preconditions.checkNotNull(lock, "Checkpoint lock should not be NULL.");
+		Preconditions.checkNotNull(operator, "Reference to AsyncWaitOperator should not be NULL.");
+
+		this.bufferSize = bufferSize;
+		this.mode = mode;
+		this.output = output;
+		this.timestampedCollector = collector;
+		this.operator = operator;
+		this.lock = lock;
+
+		this.emitter = new Emitter();
+		this.emitThread = new Thread(emitter);
+	}
+
+	/**
+	 * Add an {@link StreamRecord} into the buffer. A new {@link AsyncCollector} will be created and returned
+	 * corresponding to the input StreamRecord.
+	 * <p>
+	 * If buffer is full, caller will wait until a new space is available.
+	 *
+	 * @param record StreamRecord
+	 * @return An AsyncCollector
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	public AsyncCollector<IN, OUT> addStreamRecord(StreamRecord<IN> record) throws InterruptedException, IOException {
+		while (queue.size() >= bufferSize) {
+			// hold the input StreamRecord until it is placed in the buffer
+			extraStreamElement = record;
+
+			lock.wait();
+		}
+
+		if (error != null) {
+			throw error;
+		}
+
+		AsyncCollector<IN, OUT> collector = new AsyncCollector(this);
+
+		queue.put(collector, record);
+
+		extraStreamElement = null;
+
+		return collector;
+	}
+
+	/**
+	 * Add a {@link Watermark} into queue. A new AsyncCollector will be created and returned.
+	 * <p>
+	 * If queue is full, caller will wait here.
+	 *
+	 * @param watermark Watermark
+	 * @return AsyncCollector
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	public AsyncCollector<IN, OUT> addWatermark(Watermark watermark) throws InterruptedException, IOException {
+		return processMark(watermark);
+	}
+
+	/**
+	 * Add a {@link LatencyMarker} into queue. A new AsyncCollector will be created and returned.
+	 * <p>
+	 * If queue is full, caller will wait here.
+	 *
+	 * @param latencyMarker LatencyMarker
+	 * @return AsyncCollector
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	public AsyncCollector<IN, OUT> addLatencyMarker(LatencyMarker latencyMarker) throws InterruptedException, IOException {
+		return processMark(latencyMarker);
+	}
+
+	private AsyncCollector<IN, OUT> processMark(StreamElement mark) throws InterruptedException, IOException {
+		while (queue.size() >= bufferSize) {
+			// hold the input StreamRecord until it is placed in the buffer
+			extraStreamElement = mark;
+
+			lock.wait();
+		}
+
+		if (error != null) {
+			throw error;
+		}
+
+		AsyncCollector<IN, OUT> collector = new AsyncCollector(this, true);
```

Comment:
You could for example define a `StreamElementEntry` which can return the `StreamElement` and says whether the entry is done. Additionally you have the `AsyncCollector` interface which is implemented by a `StreamRecordEntry implements StreamElementEntry` and can be given to the `AsyncFunction`. The watermark is stored in a `WatermarkEntry implements StreamElementEntry` and the latency marker is stored in a `LatencyEntry implements StreamElementEntry`. Storing them in a `LinkedHashSet` should give you exactly the same behaviour but with an improved abstraction imho.